src/enclave/enclave_oe.c

```diff
@@ -303,20 +330,32 @@ static void _copy_shared_memory(const sgxlkl_shared_memory_t* host)
         }
     }
 
-    if (host->env)
+    /* Copy the host's environment variables to enclave memory */
+    char* const* henv = host->env;
+    size_t henvc = host->envc;
+    if (henv && henvc)
     {
-        size_t henvc = 0;
-        while (host->env[henvc++] != 0)
-            ;
+        CHECK_OUTSIDE(henv, sizeof(char*) * henvc);
         char** tmp = oe_calloc_or_die(
-            henvc + 1,
+            henvc,
             sizeof(char*),
             "Could not allocate memory for host import environment variable\n");
         for (size_t i = 0; i < henvc; i++)
-            tmp[i] = oe_strdup(host->env[i]);
-        tmp[henvc] = NULL;
+        {
+            const char* env_i = henv[i];
+            size_t n = strlen(env_i) + 1;
```

Comment:
`strlen()` operates on `env_i` buffer already, before the check of whether the buffer is outside of enclave. String buffer is hard to manage in enclave interfaces, exactly because the data marshalling code inside the enclave has to operate on the buffer to calcuate the length. If the string buffer can be defined with a max length, the marshalling code can check the the max-length string buffer is outside of enclave before calling `strlen()`.

Reply:
295 sizeof(enc_dev_config_t) * enc->evt_channel_num may overflow.

Reply:
Have you considered where the lfences should be placed?

Reply:
We don't have max sizes for all strings, e.g. this particular one pointed at here is the content of a host-controlled environment variable. I think I can pass all of the variable in one string + size and tokenize afterwards.

The overflow is fine, in that case it would reserve too little memory and segfault. I'm not putting any lfences anywhere, that will have to be a separate effort.